File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if(len(s)%2!=0):
            return False
        
        stk = []
        for i in range(len(s)):
            if(s[i]==")" or s[i]=="}" or s[i]=="]"):
                if(len(stk)!=0):
                    if(s[i]==")" and stk[-1]!="("):
                        return False
                    elif(s[i]=="]" and stk[-1]!="["):
                        return False
                    elif(s[i]=="}" and stk[-1]!="{"):
                        return False
                    else:
                        stk.pop()
                    
                else:
                    return False
            else:
                stk.append(s[i])
        
        return len(stk)==0

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if(s.length()%2!=0){
            return false;
        }
        Stack<Character> st = new Stack<>();
        for(int i = 0; i<s.length(); i++){
            if(s.charAt(i)==')' || s.charAt(i)==']' || s.charAt(i)=='}'){
                if(!st.isEmpty()){
                    if(s.charAt(i)==')' && st.peek()!='('){
                        return false;
                    }
                    if(s.charAt(i)==']' && st.peek()!='['){
                        return false;
                    }
                    if(s.charAt(i)=='}' && st.peek()!='{'){
                        return false;
                    }
                    else{
                        st.pop();
                    }
                }
                else{
                    return false;
                }
            }
            else{
                st.push(s.charAt(i));
            }
        }

        return st.isEmpty();
    }
}